main.rb

```diff
@@ -0,0 +1,186 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    return enum_for(:my_each) unless block_given?
+
+    new_self = self
+    i = 0
+    new_self = new_self.instance_of?(Array) || new_self.instance_of?(Hash) ? flatten : to_a
+    if instance_of?(Array) || instance_of?(Range)
+      new_self.length.times do
+        yield(new_self[i])
+        i += 1
+      end
+    elsif instance_of?(Hash)
+      new_self.length / 2.times do
+        yield(new_self[i], new_self[i + 1])
+        i += 2
+      end
+    end
+    new_self
+  end
```

Comment:
# my_each

- [ ] `my_each` when a block is given when self is a range returns the range itself after calling the given block once for each element in self

     Failure/Error: expect(range.my_each(&block)).to eq(range.each(&block))

       expected: 5..50
            got: [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]
##

- [ ] `my_each`, when a block is given when self is a hash, behaves like #each calls the given block once for each element in self

     Failure/Error: expect(my_each_output).to eq(each_output)

       expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: "ab"